File: cf-hostname-reconciler/src/reconciler/zones.py
"""Zone discovery and target-zone resolution."""

from __future__ import annotations

import logging
import time

from .cloudflare import CloudflareClient, CloudflareError, Zone

log = logging.getLogger(__name__)


class ZoneRegistry:
    """Caches the zone list so every reconcile does not re-list zones."""

    def __init__(
        self,
        client: CloudflareClient,
        names: tuple[str, ...] = (),
        ttl: int = 900,
    ) -> None:
        self._client = client
        self._names = names
        self._ttl = ttl
        self._zones: list[Zone] = []
        self._fetched_at = 0.0

    def zones(self, force: bool = False) -> list[Zone]:
        if force or not self._zones or time.monotonic() - self._fetched_at > self._ttl:
            zones = self._client.list_zones(self._names)
            if not zones:
                raise CloudflareError("token can see no zones; check its Zone:Read permission")
            # CF_DOMAINS order is meaningful (first = default target), so keep it.
            self._zones = zones if self._names else sorted(zones, key=lambda z: z.name)
            self._fetched_at = time.monotonic()
            log.info("zones discovered", extra={"zones": [z.name for z in self._zones]})
        return self._zones

    def zone_names(self) -> tuple[str, ...]:
        return tuple(z.name for z in self.zones())

    def by_name(self, name: str) -> Zone | None:
        name = name.strip().lower()
        return next((z for z in self.zones() if z.name == name), None)

    def target_zone(self, hint: str = "") -> Zone:
        """Zone that should own a newly created custom hostname.

        Resolution order: container `cf_zone` label -> first CF_DOMAINS entry ->
        the single known zone.
        """
        if hint:
            zone = self.by_name(hint)
            if zone:
                return zone
            raise CloudflareError(f"target zone {hint!r} is not among the managed zones")

        zones = self.zones()
        if self._names or len(zones) == 1:
            return zones[0]
        raise CloudflareError(
            "multiple zones are managed but no target zone was given; "
            "set CF_DOMAINS or add a `cf_zone` label to the container"
        )

File: cf-hostname-reconciler/src/zones.js
// Zone discovery and target-zone resolution, with a short cache.
import { CloudflareError } from "./cloudflare.js";
import { log } from "./logging.js";

export class ZoneRegistry {
  constructor(client, names = [], ttlSeconds = 900, now = () => Date.now()) {
    this._client = client;
    this._names = names;
    this._ttl = ttlSeconds * 1000;
    this._now = now;
    this._zones = [];
    this._fetchedAt = 0;
  }

  async zones(force = false) {
    if (force || this._zones.length === 0 || this._now() - this._fetchedAt > this._ttl) {
      const zones = await this._client.listZones(this._names);
      if (!zones.length) throw new CloudflareError("token can see no zones; check its Zone:Read permission");
      // CF_DOMAINS order is meaningful (first = default target); keep it. Otherwise sort.
      this._zones = this._names.length ? zones : [...zones].sort((a, b) => a.name.localeCompare(b.name));
      this._fetchedAt = this._now();
      log.info("zones discovered", { zones: this._zones.map((z) => z.name) });
    }
    return this._zones;
  }

  async zoneNames() {
    return (await this.zones()).map((z) => z.name);
  }

  async byName(name) {
    const wanted = name.trim().toLowerCase();
    return (await this.zones()).find((z) => z.name === wanted) ?? null;
  }

  // Zone that should own a newly created custom hostname.
  // Order: container `cf_zone` hint -> first CF_DOMAINS entry -> the single known zone.
  async targetZone(hint = "") {
    if (hint) {
      const zone = await this.byName(hint);
      if (zone) return zone;
      throw new CloudflareError(`target zone ${JSON.stringify(hint)} is not among the managed zones`);
    }
    const zones = await this.zones();
    if (this._names.length || zones.length === 1) return zones[0];
    throw new CloudflareError(
      "multiple zones are managed but no target zone was given; set CF_DOMAINS or add a `cf_zone` label to the container",
    );
  }
}
